File: Java Fundamentals/FinalExamPrep/src/ThePianist.java
import java.util.*;

public class ThePianist {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int numOfPieces = Integer.parseInt(scanner.nextLine());
        Map<String, List<String>> piecesMap = new LinkedHashMap<>();
        for (int i = 0; i < numOfPieces; i++) {
            String[] pieceArr = scanner.nextLine().split("\\|");
            String pieceName = pieceArr[0];
            String composer = pieceArr[1];
            String key = pieceArr[2];
            piecesMap.putIfAbsent(pieceName, new ArrayList<>());
            piecesMap.get(pieceName).add(composer);
            piecesMap.get(pieceName).add(key);
        }

        while (true) {
            String[] pieceArr = scanner.nextLine().split("\\|");
            if(pieceArr[0].equals("Stop")){break;}
            String command = pieceArr[0];
            String pieceName = pieceArr[1];

            switch (command){
                case "Add":
                    String composerName = pieceArr[2];
                    String pieceKey = pieceArr[3];
                    if(!piecesMap.containsKey(pieceName)){
                        piecesMap.put(pieceName, new ArrayList<>());
                        piecesMap.get(pieceName).add(composerName);
                        piecesMap.get(pieceName).add(pieceKey);
                        System.out.printf("%s by %s in %s added to the collection!%n", pieceName, composerName, pieceKey);
                    }else{
                        System.out.printf("%s is already in the collection!%n", pieceName);
                    }
                    break;

                case "Remove":
                    if(piecesMap.containsKey(pieceName)){
                        piecesMap.remove(pieceName);
                        System.out.printf("Successfully removed %s!%n", pieceName);
                    }else{
                        System.out.printf("Invalid operation! %s does not exist in the collection.%n", pieceName);
                    }
                    break;

                case "ChangeKey":
                    String keyNew = pieceArr[2];
                    if(piecesMap.containsKey(pieceName)){
                        piecesMap.get(pieceName).remove(piecesMap.get(pieceName).size() - 1);
                        piecesMap.get(pieceName).add(keyNew);
                        System.out.printf("Changed the key of %s to %s!%n", pieceName, keyNew);
                    }else{
                        System.out.printf("Invalid operation! %s does not exist in the collection.%n", pieceName);
                    }
                    break;

                default:
                    break;
            }
        }

        for(Map.Entry<String, List<String>> entry : piecesMap.entrySet()){
            String name = entry.getKey();
            String composer = entry.getValue().get(0);
            String key = entry.getValue().get(1);
            System.out.printf("%s -> Composer: %s, Key: %s%n", name, composer, key);
        }
    }
}

File: JS Front-End/Exam Prep/Prep_1/ThePianist.js
function solve(input){
    let countOfPieces = Number(input[0]);
    let piecesMap = {};

    let i = 1;
    while(i <= countOfPieces){
        let pieceInfo = input[i].split("|");
        let pieceName = pieceInfo[0];
        let composer = pieceInfo[1];
        let key = pieceInfo[2];
        piecesMap[pieceName] = {composer, key};

        i++;
    }

    let currentLine = input[i];
    while(currentLine != "Stop"){
        let request = currentLine.split("|");
        let pieceName = request[1];

        switch(request[0]){
            case "Add":
                if(!piecesMap.hasOwnProperty(pieceName)){
                    let composer = request[2];
                    let key = request[3];
                    piecesMap[pieceName] = {composer, key};
                    console.log(`${pieceName} by ${composer} in ${key} added to the collection!`);
                }else{
                    console.log(`${pieceName} is already in the collection!`)
                }
                break;
            case "Remove":
                if(piecesMap.hasOwnProperty(pieceName)){
                    delete piecesMap[pieceName];
                    console.log(`Successfully removed ${pieceName}!`);
                }else{
                    console.log(`Invalid operation! ${pieceName} does not exist in the collection.`)
                }
                break;
            case "ChangeKey":
                if(piecesMap.hasOwnProperty(pieceName)){
                    let composer = piecesMap[pieceName].composer;
                    let key = request[2];
                    piecesMap[pieceName] = {composer, key};
                    console.log(`Changed the key of ${pieceName} to ${key}!`);
                }else{
                    console.log(`Invalid operation! ${pieceName} does not exist in the collection.`)
                }
                break;
        }

        i++;
        currentLine = input[i];
    }

    let entries = Object.entries(piecesMap);
    for (const [ piece, info ] of entries) {
        console.log(`${piece} -> Composer: ${info.composer}, Key: ${info.key}`);
    }
}

solve(
    [
        '3',
        'Fur Elise|Beethoven|A Minor',
        'Moonlight Sonata|Beethoven|C# Minor',
        'Clair de Lune|Debussy|C# Minor',
        'Add|Sonata No.2|Chopin|B Minor',
        'Add|Hungarian Rhapsody No.2|Liszt|C# Minor',
        'Add|Fur Elise|Beethoven|C# Minor',
        'Remove|Clair de Lune',
        'ChangeKey|Moonlight Sonata|C# Major',
        'Stop'  
      ]      
);
console.log("------------------------------------------------------------")
solve(
    [
        '4',
        'Eine kleine Nachtmusik|Mozart|G Major',
        'La Campanella|Liszt|G# Minor',
        'The Marriage of Figaro|Mozart|G Major',
        'Hungarian Dance No.5|Brahms|G Minor',
        'Add|Spring|Vivaldi|E Major',
        'Remove|The Marriage of Figaro',
        'Remove|Turkish March',
        'ChangeKey|Spring|C Major',
        'Add|Nocturne|Chopin|C# Minor',
        'Stop'
      ]      
);